### PR TITLE
feat: support DTensor CP in DPO and SFT

### PR DESCRIPTION
**Pull request by ashors1**
# What does this PR do ?
Enables CP with DTensor for DPO and SFT

DPO loss: 
dpo_cp

SFT loss:
sft_cp

# Issues
List issues that this PR closes ([syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)):

# Usage
* **You can potentially add a usage example below**

```python
# Add a code snippet demonstrating how to use this
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

### CODE REVIEW

File: nemo_rl/algorithms/loss_functions.py [hunk 1]
```diff
@@ -325,6 +325,7 @@ def __call__(
         token_mask = data["token_mask"][:, 1:]
         sample_mask = data["sample_mask"]
         mask = token_mask * sample_mask.unsqueeze(-1)
+        seq_index = data.get("seq_index", None)
```
Comment (terrykong) [inline nemo_rl/algorithms/loss_functions.py:328]: should this be not `None` if cp is enabled? if so, can we assert?

Comment (joyang-nv) [reply nemo_rl/algorithms/loss_functions.py:328]: Yes, agreed.

Comment (ashors1) [reply nemo_rl/algorithms/loss_functions.py:328]: We already have a similar assertion here: https://github.com/NVIDIA-NeMo/RL/blob/main/nemo_rl/distributed/model_utils.py#L182. Do you think we should add another assertion to the loss functions themselves?

Comment (terrykong) [reply nemo_rl/algorithms/loss_functions.py:328]: Yea, it's more clear when you look at this function what it expects

**Review by terrykong — APPROVED** [review]

File: nemo_rl/algorithms/loss_functions.py [hunk 2]
```diff
@@ -155,8 +155,22 @@ def __call__(
...
+
+            cp_size = 1
+            if (
+                target.device_mesh.mesh_dim_names is not None
+                and "cp" in target.device_mesh.mesh_dim_names
+            ):
+                cp_dim_index = target.device_mesh.mesh_dim_names.index("cp")
+                cp_size = target.device_mesh.shape[cp_dim_index]
+            if cp_size > 1:
+                assert seq_index is not None, (
```
Comment (terrykong) [inline nemo_rl/algorithms/loss_functions.py]: actually i just realized, can you do all of this assertion inside `get_logprobs_from_vocab_parallel_logits` since there's already an assertion there? so check if `cp > 1` then assert seq_index can't be None

prob would simplify these checks that occur in multiple places

Comment (ashors1) [reply nemo_rl/algorithms/loss_functions.py]: I can do that, but imo it feels a bit redundant since the same check is done one level down in `dtensor_from_parallel_logits_to_logprobs`. Do you think we should move the check from `dtensor_from_parallel_logits_to_logprobs` to `get_logprobs_from_vocab_parallel_logits`?

Comment (ashors1) [reply nemo_rl/algorithms/loss_functions.py]: In an offline discussion, we decided the existing assertion in `dtensor_from_parallel_logits_to_logprobs` is sufficient
